src/challenge.rs

```diff
@@ -0,0 +1,57 @@
+use ark_ff::PrimeField;
+use ark_sponge::FieldBasedCryptographicSponge;
+
+/// State stored for univariate generator
+pub struct UnivariateGeneratorState<F: PrimeField> {
+    gen: F,
+    next: F,
+}
+
+impl<F: PrimeField> UnivariateGeneratorState<F> {
+    fn new(gen: F) -> Self {
+        Self { gen, next: gen }
+    }
+
+    fn get_next(&mut self) -> F {
+        let result = self.next;
+        self.next *= self.gen;
+        result
+    }
+}
+
+// TODO: Copy and Clone trait cannot be derived
```

Comment:
maybe we need to first derive the Copy and Clone trait for UnivariateGeneratorState above, which should be doable.

Reply:
I remembered what the compiler complained is that `&mut Sponge` does not have Copy and Clone trait

Reply:
fixed and mark as resolved